### PR TITLE
DS-3513 Provide an entry endpoint that exposes the known endpoints via the HAL browser

### PR DESCRIPTION
**Pull request by abollini**
Hi @terrywbrady 
here my initial implementation of an endpoint that exposes the known endpoints via the HAL browser.
Start from
https://github.com/4Science/DSpace/blob/DS-3513/dspace-spring-rest/src/main/java/org/dspace/app/rest/RootRestResourceController.java

There are several area that needs improvements
https://github.com/4Science/DSpace/blob/DS-3513/dspace-spring-rest/src/main/java/org/dspace/app/rest/DiscoverableEndpointsService.java#L81
https://github.com/4Science/DSpace/blob/DS-3513/dspace-spring-rest/src/main/java/org/dspace/app/rest/DiscoverableEndpointsService.java#L68
https://github.com/4Science/DSpace/blob/DS-3513/dspace-spring-rest/src/main/java/org/dspace/app/rest/RestResourceController.java#L67

### CODE REVIEW

File: dspace-spring-rest/src/main/java/org/dspace/app/rest/DiscoverableEndpointsService.java [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+	/**
+	 * Is the computed list of discoverableEndpoints valid?
+	 */
+	private boolean initialized = false;
+
+	private List<Link> discoverableEndpoints = new ArrayList<Link>();
+
+	public void register(Object controller, List<Link> links) {
+		synchronized (this) {
+			initialized = false;
```
Comment (terrywbrady) [inline dspace-spring-rest/src/main/java/org/dspace/app/rest/DiscoverableEndpointsService.java]: Should this be set to true?

**Review by terrywbrady — APPROVED** [review]
Tested locally.  This works very nicely, and it is good to see the approach you took.

A couple additional thoughts. 
- Should all of the other endpoints have a link back to this endpoint for convenience?
- Would it be simpler to rebuild the list of links on every request rather than having the synchronized methods?  How often will this endpoint actually be invoked?

File: dspace-spring-rest/src/main/java/org/dspace/app/rest/DiscoverableEndpointsService.java [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+						// sanity check
+						// FIXME improve logging for debugging
+						if (rels.contains(link.getRel())) {
+							throw new IllegalStateException("The rel " + link.getRel() + " is defined multiple times!");
+						}
+						rels.add(link.getRel());
+					}
+				}
+			}
+
```
Comment (abollini) [inline dspace-spring-rest/src/main/java/org/dspace/app/rest/DiscoverableEndpointsService.java]: now we can set

`initialized = true;`

File: dspace-spring-rest/src/main/java/org/dspace/app/rest/DiscoverableEndpointsService.java (hunk 1 above)
Comment (abollini) [inline dspace-spring-rest/src/main/java/org/dspace/app/rest/DiscoverableEndpointsService.java]: no, it should be done after the creation of the result lists in the getDiscoverableEndpoints method (see comment below)

**Comment by abollini** [comment]
> Should all of the other endpoints have a link back to this endpoint for convenience?
This is typically not done. Spring Data Rest doesn't do that, it is arguable but a client should start to use the API from the ROOT so this is an external know information, there are no value to add this information in all the response making them larger than necessary

> Would it be simpler to rebuild the list of links on every request rather than having the synchronized methods? How often will this endpoint actually be invoked?
Following a HAL approach, each client should make this initial request to register which endpoints are needed for each specific functionality. The client should not assume that /api/core/items is the endpoint to use to deal with items, it should only know that our API expose a items resource, the actual link is what is expressed in the ROOT document. This should allow easy replacement of an endpoint with a local customized version.

Compute the links could be intensive once we deal with overriding and customization, this is way I have introduced a cache / syncronization

**Pull request merged**